[clang-tidy] Improve alternate snake case warnings

jcmoyer opened the pull request:
Improves the accuracy of `readability-identifier-naming` for cases `Camel_Snake_Case` and `camel_Snake_Back`. Prior to this commit, these cases matched identifiers with **only** a leading upper case letter or leading lower case letter respectively. Now, uppercase letters can only appear at the start of an identifier or directly following an underscore.

---

Currently, the regex for `Camel_Snake_Case` matches any identifier that starts with a capital letter:

```
^[A-Z]([a-z0-9]*(_[A-Z])?)*
                ^^^^^^^^^-- underscore + capital letter after the first capital is optional
```

This means that `Camel_Snake_Case` matches other cases - in particular `CamelCase` and `Leading_upper_snake_case` - which causes clang-tidy to sometimes not flag incorrect casing. It also matches `UPPER_CASE`, but I think it's reasonable to consider this a subset of `Camel_Snake_Case` since some users may prefer e.g. `XML_Parser` to `Xml_Parser`. It's really easy to accidentally type an identifier that clang-tidy doesn't catch; all you have to do is omit an underscore or forget to capitalize a letter. The same problem also applies to `camel_Snake_Back` except that any identifier starting with a lower case letter matches, so I went ahead and adjusted its regex too. Fixing it also uncovered a minor error in an existing test.

Here are some examples of the current behavior when using `Camel_Snake_Case`:

```cpp
// accepted by clang-tidy
String_Reader
StringReader  // bad, CamelCase
String_reader // bad, Leading_upper_snake_case
Xml_Parser_2
XML_Parser_2
XML_PaRser_2  // bad? mixed case in second word
T
Foo
A_B_C
A_b_c         // bad, Leading_upper_snake_case

// rejected by clang-tidy
t
foo
a_b_c
```

After this patch:

```cpp
// accepted by clang-tidy
String_Reader
Xml_Parser_2
XML_Parser_2
T
Foo
A_B_C

// rejected by clang-tidy
StringReader
String_reader
XML_PaRser_2
t
foo
A_b_c
a_b_c
```

PiotrZSL commented on clang-tools-extra/clang-tidy/readability/IdentifierNamingCheck.cpp:
```diff
@@ -871,8 +871,8 @@ bool IdentifierNamingCheck::matchesStyle(
-      llvm::Regex("^[a-z]([a-z0-9]*(_[A-Z])?)*"),
+      llvm::Regex("^[A-Z]+([a-z0-9]*_[A-Z0-9]+)*[a-z0-9]*$"),
+      llvm::Regex("^[a-z]+([a-z0-9]*_[A-Z0-9]+)*[a-z0-9]*$"),
       llvm::Regex("^[A-Z]([a-z0-9_]*[a-z])*$"),
```
Shouldn't this LeadingUpperSnakeCase also be changed ?

jcmoyer replied:
I think it's fine, `Leading_upper_snake_case` only has the first letter capitalized and no special requirements beyond that. Though maybe the regex could be slightly simplified to something like `^[A-Z][a-z0-9_]*[a-z]$`.

PiotrZSL reviewed (approved):
LGTM

PiotrZSL commented:
Also update release notes, add one sentence there about this.

jcmoyer commented:
> Also update release notes, add one sentence there about this.

Done.

jcmoyer commented:
Rebased to resolve merge conflict.

PiotrZSL commented:
@jcmoyer Do you want this be merged as jcmoyer@users.noreply.github.com ? If not then please disable privacy settings in GitHub or change author of commit.

jcmoyer committed:
[clang-tidy] Improve alternate snake case warnings

Improves the accuracy of `readability-identifier-naming` for cases
`Camel_Snake_Case` and `camel_Snake_Back`. Prior to this commit, these
cases matched identifiers with only a leading upper case letter or
leading lower case letter respectively. Now, uppercase letters can only
appear at the start of an identifier or directly following an
underscore.

jcmoyer commented:
@PiotrZSL I don't mind but I changed it just in case it's an issue.
